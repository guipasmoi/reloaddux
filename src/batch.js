/**
 * Created by guipa on 26/05/2017.
 */
export const batchType = "@@reloaddux/batch";

export function isBatchAction(action) {
  return action && typeof action === "object" && action[batchType] === true;
}

export function createBatch(type, ...actions) {
  return {
    type,
    [batchType]: true,
    actions
  };
}

export function wrapReducer(reducer) {
  return (state, action) => {
    if (isBatchAction(action)) {
      return action.actions.reduce(reducer, state);
    }
    return reducer(state, action);
  };
}
