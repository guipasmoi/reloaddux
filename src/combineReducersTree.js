/* eslint no-param-reassign: "off" */
import isArray from "lodash/isArray";
import isPlainObject from "lodash/isPlainObject";
import merge from "lodash/merge";
import set from "lodash/set";

function isValidNode(node) {
  return isPlainObject(node);
}

function isActionsArray(actions) {
  return actions === undefined ||
    (isArray(actions) &&
      actions.reduce(
        (accu, action) => accu && typeof action === "string",
        true
      ));
}

function isLeaf(node) {
  return node &&
    typeof node === "object" &&
    isActionsArray(node.actions) &&
    typeof node.reducer === "function";
}

export const initAction = { type: "@@redux/INIT" };

export default function combineReducersTree(
  tree,
  initAction = { type: "@@redux/INIT" }
) {
  const reversedTree = {
    // action1: { scope1: reducer, scope2: reducer },
    "*": {}
  };

  let hasBeenInitialized = false;

  function reverseTree(subTree, scope = "") {
    Object.entries(subTree).forEach(([key, node]) => {
      if (isLeaf(subTree)) {
        if (subTree.actions) {
          subTree.actions.forEach(action => {
            if (reversedTree[action] === undefined) {
              reversedTree[action] = {};
            }
            set(reversedTree[action], scope, subTree);
          });
        } else {
          set(reversedTree["*"], scope, subTree);
        }
      } else if (isValidNode(subTree)) {
        reverseTree(node, scope.length === 0 ? key : `${scope}.${key}`);
      }
    });
  }

  reverseTree(tree);
  // return reversedTree;
  function recursiveProcess(state, action, task) {
    if (isLeaf(task)) {
      if (
        task.actions === undefined || task.actions.indexOf(action.type) >= 0
      ) {
        const newState = task.reducer(state, action);
        return { value: newState, hasChanged: newState !== state };
      }
      return { value: undefined, hasChanged: undefined !== state };
    } else if (!isPlainObject(task)) {
      // reversedTree (task) must contain only plain object and reducer
      throw new Error("unexpected");
    }
    const resultValue = Object.entries(task).reduce(
      (accu, [key, subtask]) => {
        const temp = recursiveProcess(
          isPlainObject(state) ? state[key] : undefined,
          action,
          subtask
        );
        accu.value[key] = temp.value;
        accu.hasChanged = accu.hasChanged || temp.hasChanged;
        return accu;
      },
      { value: { ...state }, hasChanged: false }
    );
    return {
      ...resultValue,
      hasChanged: resultValue.hasChanged || resultValue.value !== state
    };
  }

  // todo optmization no avoid merge on each action
  return (state, action) => {
    let temp = { value: state, hasChanged: false };
    if (!hasBeenInitialized || initAction.type === action.type) {
      temp = recursiveProcess(state, action, tree);
      hasBeenInitialized = true;
    } else {
      temp = recursiveProcess(
        temp.value,
        action,
        merge({}, reversedTree[action.type], reversedTree["*"])
      );
    }
    return temp.hasChanged ? temp.value : state;
  };
}
