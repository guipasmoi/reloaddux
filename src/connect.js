/* eslint-disable no-unused-vars */
import React, { Component } from "react";
import { Connect } from "react-redux";

export default function(business = {}) {
  const {
    sagas = [],
    reducerTrees = [],
    mapStateToProps,
    mapDispatchToProps
  } = business;

  const shouldUseConnect = typeof mapDispatchToProps === "function" ||
    typeof mapStateToProps === "object";
  const shouldUseBusiness = sagas.length > 0 || reducerTrees.length > 0;

  // connector is used only when shouldUseBusiness
  const connector = shouldUseBusiness
    ? Connect(mapStateToProps, mapDispatchToProps)
    : null;

  return ComponentToWrapped => {
    if (!shouldUseBusiness && !shouldUseConnect) {
      return ComponentToWrapped;
    }
    const ConnectedComponent = connector(ComponentToWrapped);
    if (!shouldUseBusiness) {
      return ConnectedComponent;
    }
    return class extends Component {
      static contextTypes = {
        store: React.PropTypes.object
      };

      constructor(props, context) {
        super(props, context);
        this.store = this.context.store;
      }

      componentWillMount() {
        this.store.registerBusiness(business);
      }

      componentWillUnmount() {
        this.store.unregisterBusiness(business);
      }

      render() {
        return shouldUseConnect
          ? <ConnectedComponent {...this.props} />
          : <ComponentToWrapped {...this.props} />;
      }
    };
  };
}
