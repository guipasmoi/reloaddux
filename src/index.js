// @flow
export { default as withBusiness } from "./withBusiness";

export { default as Store } from "./Store";
