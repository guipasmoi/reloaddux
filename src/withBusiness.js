/* eslint-disable no-unused-vars */
import React, { Component } from "react";

export default function withBusiness(business) {
  return ComponentToWrapped =>
    class extends Component {
      static contextTypes = {
        store: React.PropTypes.object
      };

      constructor(props, context) {
        super(props, context);
        this.store = this.context.store;
      }

      componentWillMount() {
        this.store.registerBusiness(business);
      }

      componentWillUnmount() {
        this.store.unregisterBusiness(business);
      }

      render() {
        return <ComponentToWrapped {...this.props} />;
      }
    };
}
