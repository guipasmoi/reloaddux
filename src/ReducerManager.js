import merge from "lodash/merge";
import without from "lodash/without";
import uniq from "lodash/uniq";
import combineReducersTree from "./combineReducersTree";

export default class ReducerManager {
  scopeReducerMap = {};
  reducer = (action, state) => this.trueReducer(action, state);
  reducersTreeArray = [];
  trueReducer = (action, state) => state;

  addReducersTree(...trees) {
    this.reducersTreeArray = uniq(this.reducersTreeArray.concat(trees));
    this.trueReducer = combineReducersTree(
      merge({}, ...this.reducersTreeArray)
    );
  }

  removeReducersTree(...trees) {
    this.reducersTreeArray = without(this.reducersTreeArray, ...trees);
    this.trueReducer = combineReducersTree(
      merge({}, ...this.reducersTreeArray)
    );
  }
}
