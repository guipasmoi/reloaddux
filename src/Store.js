// redux
import { applyMiddleware, compose, createStore } from "redux";
import createSagaMiddleware, { runSaga } from "redux-saga";
import ReducerManager from "./ReducerManager";
import { initAction } from "./combineReducersTree";
import { wrapReducer, wrapCallBackNotification } from "./batch";

export default class Store {
  constructor(
    {
      preloadedState,
      sagaMonitor,
      middlewares = [],
      composeEnhancer = compose
    } = {}
  ) {
    const sagaMiddleware = createSagaMiddleware({ sagaMonitor });
    const allMiddlewares = [sagaMiddleware, ...middlewares];

    const reducerManager = new ReducerManager();
    this.sagaTotasksMap = new Map();

    // eslint-disable-next-line no-unused-vars
    const store = createStore(
      wrapReducer(reducerManager.reducer),
      preloadedState,
      composeEnhancer(
        applyMiddleware(...allMiddlewares)
        // other store enhancers if any
      )
    );

    const listeners = new Map();
    const notifyAll = action =>
      listeners.forEach(cb => wrapCallBackNotification(cb)(action));
    const IO = {
      subscribe: cb => {
        const token = {};
        listeners.set(token, action => cb(action));
        return () => listeners.delete(token);
      },
      dispatch: action => {
        store.dispatch(action);
        notifyAll(action);
      },
      getState: (...args) => store.getState(...args)
    };

    // the current instance become the store we just created (we added run saga fct)
    Object.assign(this, store, {
      run: sagaMiddleware.run,
      reducerManager,
      IO,
      dispatch: IO.dispatch
    });
  }

  registerBusiness(business) {
    this.reducerManager.addReducersTrees(business.reducersTree);
    Object.entries(business.sagasMap).forEach(([key, saga]) => {
      this.sagaTotasksMap.set(key, runSaga(saga(), this.IO));
    });
    this.dispatch(initAction);
  }

  unregisterBusiness(business) {
    this.reducerManager.removeReducersTree(...business.reducersTree);
    business.sagasMap.forEach(saga =>
      this.sagaTotasksMap.delete(this.run(saga)));
  }
}
