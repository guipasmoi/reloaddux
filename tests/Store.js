import React, { Component } from "react";
import { put, take } from "redux-saga/effects";
import { connect, Store } from "../src/index";

const counter1 = state => state + 1;
const counter2 = (state = 11) => state + 2;

const ACTION1 = "ACTION1";
const ACTION2 = "ACTION2";
const ACTION3 = "ACTION3";

const reducersTree1 = {
  AwesomeStateForCounter1: {
    actions: [ACTION1],
    reducer: counter1,
    default: 11
  }
};

const reducersTree2 = {
  Awesome: {
    StateForCounter2: {
      actions: [ACTION3],
      reducer: counter2
    }
  }
};

function action1(param) {
  return { type: ACTION1, param };
}

function* saga1() {
  // yield put(action1("some data awesome"));
}

function* saga2() {
  yield take(ACTION2);
  yield put(action1("some data awesome"));
}

const business1 = {
  reducersTree: reducersTree1,
  sagasMap: {
    saga1: saga1()
  },
  mapStateToProps: state => state,
  mapDispatchToProps: { action1 }
};

const business2 = {
  reducersTree: reducersTree2,
  sagasMap: {
    saga1: saga2()
  }
};

describe("Store", () => {
  it("it can create a store", () => {
    const store = new Store();
    expect(store).toMatchSnapshot();
  });
  it("it can create store and dispatch action", () => {
    // 1
    const store = new Store();
    store.dispatch({ type: ACTION1 });
    expect(store.getState()).toMatchSnapshot();

    // 2
    //store.registerBusiness(business1);
    store.dispatch({ type: "registerBusiness", business: business1 });
    expect(store.getState()).toMatchSnapshot();

    // 3
    store.dispatch({ type: ACTION1 });
    expect(store.getState()).toMatchSnapshot();

    // 4
    store.registerBusiness(business2);
    expect(store.getState()).toMatchSnapshot();

    // 5
    store.dispatch({ type: ACTION1 });
    expect(store.getState()).toMatchSnapshot();

    // 6
    store.dispatch({ type: ACTION2 });
    expect(store.getState()).toMatchSnapshot();
  });
});
