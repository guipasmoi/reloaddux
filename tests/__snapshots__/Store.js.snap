// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store it can create a store 1`] = `
Store {
  "IO": Object {
    "dispatch": [Function],
    "getState": [Function],
    "subscribe": [Function],
  },
  "dispatch": [Function],
  "getState": [Function],
  "reducerManager": ReducerManager {
    "reducer": [Function],
    "reducersTreeArray": Array [],
    "scopeReducerMap": Object {},
    "trueReducer": [Function],
  },
  "replaceReducer": [Function],
  "run": [Function],
  "sagaTotasksMap": Array [],
  "subscribe": [Function],
  Symbol(Symbol.observable): [Function],
}
`;

exports[`Store it can create store and dispatch action 1`] = `Object {}`;

exports[`Store it can create store and dispatch action 2`] = `Object {}`;

exports[`Store it can create store and dispatch action 3`] = `
Object {
  "AwesomeStateForCounter1": 12,
}
`;

exports[`Store it can create store and dispatch action 4`] = `
Object {
  "AwesomeStateForCounter1": 12,
}
`;

exports[`Store it can create store and dispatch action 5`] = `
Object {
  "Awesome": Object {
    "StateForCounter2": undefined,
  },
  "AwesomeStateForCounter1": 13,
}
`;

exports[`Store it can create store and dispatch action 6`] = `
Object {
  "Awesome": Object {
    "StateForCounter2": undefined,
  },
  "AwesomeStateForCounter1": 14,
}
`;
